### ISSUE TITLE
Suggestion: use `yield.sent` instead of `function.sent`

### ISSUE THREAD

**Issue by dead-claudia** [tc39-transfer/proposal-function.sent#1]
It IMHO explains itself a little better, and it looks more generator specific than `function.sent` (which looks like it would belong in any *function*, not just any generator).

**Comment by hax** [comment]
All names of `yield.xxx` suffer from the issues:
1.  `yield.xxx` and `(yield).xxx` have different semantics in generators. 
1. `yield.xxx` have different semantics in generators (meta property) and non-strict non-generator codes (normal property).
1. `yield.xxx` and `yield xxx` looks close if the font of `.` is too inconspicuous 😂
1. `yield.xxx` and `yield*xxx` looks close if the font of `.` is too conspicuous 🤣 
1. `yield.xxx` implies something caused by `yield` but actually the main reason of this proposal is about the first sent value before any `yield` (or without any `yield`).

I feel no any single issue is fatal, but there are too many potential ambiguity so maybe we should first consider alternatives of `function.xxx` instead of `yield.xxx`.

**Comment by ljharb** [comment]
I don't think we have to worry about the font; anything can be unreadable with the wrong font.

Adding parens to things causes different semantics all over the place; i don't think that's a concern either - similarly, `yield` has a different semantic in generators vs non generators, so that's also not a concern.

However, your last point is compelling; it'd be great to find another alternative.

**Comment by hax** [comment]
> Adding parens to things causes different semantics all over the place

@ljharb  Could you explain more about that? I don't know other case which have different semantics between `identifier.xxx` with `(identifier).xxx`

> it'd be great to find another alternative

Do you have any options? 😀

**Comment by ljharb** [comment]
- `(a?.b).c` vs `a?.b.c`
 - `a, b.c` vs `(a, b).c`
 - `a + b + c` vs `(a + b) + c` vs `a + (b + c)` when all three are different types
etc

No suggestions yet.

**Comment by hax** [comment]
@ljharb What I mean is member access (or similar, like metaproperty) chaining 😅

> - `a, b.c` vs `(a, b).c`
> - `a + b + c` vs `(a + b) + c` vs `a + (b + c)`

So I don't think these two cases are comparable to `yield.xxx` vs `(yield).xxx` at all.

>  - `(a?.b).c` vs `a?.b.c`

Actually this case [seems weird](https://github.com/tc39/proposal-optional-chaining/issues/33#issuecomment-331536970) at first glance, and there were [several](https://github.com/tc39/proposal-optional-chaining/issues/2) [issues](https://github.com/tc39/proposal-optional-chaining/issues/10) and [discussions](https://github.com/tc39/proposal-optional-chaining/issues/33) about that. IMO `yield.xxx` vs `(yield).xxx` is even worse because: it seems no real use case of `(a?.b).c`, but `(yield).xxx` in generator has valid use cases.

**Comment by hax** [comment]
I have some thought about the name.

The real meaning of `function.sent` in a generator function is "the value **sent** by the last `next()` call to a generator object which created by this generator **function**". The problem is, strictly speaking, the value is sent to "generator object" not "generator function", so `function.sent` sounds more like "a function instance sent to somewhere".

I feel if we use `function.xxx` it should sound like "the value that generator function got".

Here are the names go to my mind:

- `function.got`
- `function.input`
- `function.lastValue`
- `function.currentValue`
- `function.receivedValue`
- `function.received`
- `var.received` (while `var` seems have no relationship with generator, it could be understand as a special variable in the generator function)

**Comment by dead-claudia** [comment]
@hax What about `function.received`?

**Comment by hax** [comment]
@isiahmeadows Oh, I forgot to add it. Added now.

Currently I slightly prefer `function.got` or `function.input` because they are shorter 😛. 

I also slightly worry about `function.received` could be considered to be a boolean.

What do you think of these names?

**Comment by ljharb** [comment]
I remain concerned about using `function` for something that doesn't apply to all functions :-/

**Comment by ljharb** [comment]
maybe `function*.sent` would work though (swap out "sent" for whatever the bikeshed produces)

**Comment by hax** [comment]
Though `function*.sent` may work, it looks very unlike a "meta **property**", not sure how committee would accept it 😅 . And addition of `*` also make ergonomics a little bit worse (`function` is already long enough)

**Comment by lifaon74** [comment]
For me the best would be something able to retrieve every values sent though `next` (not only the first one)

`keyWord` may be any of: `last`, `sent`, `received`, `input`, ... => TO BE DEFINED

I agree with @ljharb , `function.keyWord` may be a little confusing. This specs only apply to generator functions intead of everyone.

`(yield).keyWord` => resolves first the yield, then go to the 'keyWord' property
`yield.keyWord` => returns the last value provided to `next`

I personally prefer `yield.received` as its for me the most explicit.

**Comment by dead-claudia** [comment]
Note: `function *f() { yield.received }` throws an early error in Chrome, at least.

I do agree with @allenwb that `yield.anything` is not ideal, since it's a valid identifier anywhere else.

**Comment by Jamesernator** [comment]
> Note: `function *f() { yield.received }` throws an early error in Chrome, at least.
> 
> I do agree with @allenwb that `yield.anything` is not ideal, since it's a valid identifier anywhere else.

I'm not convinced that the fact `yield` is an identifier in sloppy mode is that compelling of a reason not to use it. It already has different meanings depending on context for tagged template literals (any others?) `` yield `foobar`; ``,

Although I do think the difference between `(yield).blah` vs `yield.blah` might be slightly confusing though given they can both occur in the same context. However I do feel like people wouldn't use `(yield).prop` as much with `yield.blah.prop` available anyway.

**Comment by hax** [comment]
I am considering maybe `lastValue@yield` could work 😂 though it still not solve:

> - `yield.xxx` (or `xxx@yield`)  implies something caused by `yield` but actually the main reason of this proposal is about the first sent value before any `yield` (or without any `yield`).

**Comment by Jack-Works** [comment]
I strongly agree with `yield.received`

**Comment by jimmywarting** [comment]
I strongly disagree with `function.sent` think something other would be better
